iterators.tex

```diff
@@ -623,31 +650,186 @@
+\range{i}{j} denotes a range. Let $N$ be the smallest number of
+applications of \tcode{++i} necessary to make \tcode{bool(i == j)} be
+\tcode{true}. \tcode{SizedIterator<I>()} is satisfied if and only if:
+
+\begin{itemize}
+\item If $N$ is representable by \tcode{difference_type_t<I>},
+      then \tcode{j - i} is well-defined and equals $N$.
+
+\item If $-N$ is representable by \tcode{difference_type_t<I>},
+      then \tcode{i - j} is well-defined and equals $-N$.
+\end{itemize}
+\end{itemdescr}
+
+\pnum
+The \tcode{disable_sized} type predicate provides a mechanism to enable
+use of iterator types that meet the syntactic requirements but do not
+in fact satisfy \tcode{SizedIterator} with the library.
 
 \pnum
-The \tcode{Sentinel} concept defines requirements for a type that
+\enternote A program that instantiates a library template that requires
+an \tcode{Iterator} with such an iterator type \tcode{I} is ill-formed
+with no diagnostic required unless \tcode{disable_sized<I>} evalutes to
+\tcode{true}~(\ref{structure.requirements}). \exitnote
+
+\enternote The \tcode{SizedIterator} concept is satisfied by
+\tcode{RandomAccessIterator}s~(\ref{iterators.random.access}) and by
+counted iterators~(\ref{counted.iterator}).\exitnote
+
+\ednote{This concept also gives us a way to demote the category of \tcode{move_iterator}s to Input
+while retaining the ability of \tcode{move_iterator} pairs to communicate the range's size to
+container constructors.}
+} %% color{newclr}
+
+\rSec2[iterators.sentinel]{Concept Sentinel}
+\pnum
+The \tcode{Sentinel} concept \oldtxt{defines requirements for a type that
 is an abstraction of the past-the-end iterator. Its values can be
-compared to an iterator for equality.
+compared to an iterator for equality.}\newtxt{specifies requirements on
+the types \tcode{S} and \tcode{I} of a sentinel and an iterator that
+denote a range.}
 
 \indexlibrary{\idxcode{Sentinel}}%
-\begin{codeblock}
-  template <class T, class I>
+\begin{itemdecl}
+  template <class @\oldtxt{T}\newtxt{S}@, class I>
   concept bool Sentinel() {
-    return Regular<T>() &&
+    return Regular<@\oldtxt{T}\newtxt{S}@>() &&
       Iterator<I>() &&
-      EqualityComparable<T, I>();
+      EqualityComparable<@\oldtxt{T}\newtxt{S}@, I>();
   }
-\end{codeblock}
+\end{itemdecl}
+
+{\color{newclr}
+\begin{itemdescr}
+Let \tcode{s} and \tcode{t} be values of type \tcode{S},
+and \tcode{i} a value of type \tcode{I} such that \range{i}{s}
+denotes a range. Types \tcode{S} and \tcode{I} satisfy
+\tcode{Sentinel<S, I>()} if and only if:
+
+\begin{itemize}
+\item \tcode{i == s}, \tcode{s == i},\tcode{i != s}, and
+      \tcode{s != i} are well-defined.
+
+\item \tcode{bool(i == s)} if and only if \tcode{i} and
+      \tcode{s} denote the same element.
+
+\item If \range{i}{s} and \range{i}{t} denote the same range,
+      then \tcode{s == t} is well-defined and \tcode{s}
+      equals \tcode{t}.
```

Comment:
>
> Yes, that's the gist of it. Sentinels can be Regular, but they cannot be
> stateful predicates and simultaneously model cross-type EqualityComparable
> with an Iterator. Cross-type EC requires that whenever one iterator is ==
> two sentinels, those sentinels are equal.
>
Ah... Of course that's where this is coming from.

When Marcin and I were first writing the Palo Alto paper, we discussed
whether or not the cross-type concepts should require the independent types
to be EC. What appears in the report is what was discussed during the
meeting, but there were some good reasons for relaxing the requirement.

I would say this: even if EqualityComparable<T,U> is satisfied, and for
some t and u, t==u, there is no guarantee that t can be s substituted for u
in all EP expressions. For example, a call f(t) might be well formed, but
f(u) might not --- or worse, you might find an overload with different
semantics than f(t).

This strikes me as one of those cases where i == s just means that i can be
compared for equality with s, but where that capability should impose no
other requirements on the types of i or s.


> This axiom is actually a consequence of EC for finite ranges. I added it
> as an axiom to enable infinite ranges to satisfy EC since one can't
> presuppose the existence of an iterator equal to the sentinel in an
> infinite range.
>
I don't think this is necessary. Since all finite ranges are expected
(required?) to be EC, you should be able to cover them with a semantic
requirement on the concept as a whole.


All finite ranges satisfy are EqualityComparable. For all ranges r1 and r2
(same type), r1 == r2 is true iff r1 and r2 cover (denote? range over?,
contain?) the same elements. That is, iterating over both ranges produces
the same sequence of objects.

That should be sufficient to impose implementation requirements on the
sentinel without precisely saying how it is achieved.

Andrew